File: user-service/src/main/java/rs/raf/user_service/service/CompanyService.java
package rs.raf.user_service.service;

import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;
import rs.raf.user_service.domain.dto.CompanyDto;
import rs.raf.user_service.domain.dto.CreateCompanyDto;
import rs.raf.user_service.domain.entity.ActivityCode;
import rs.raf.user_service.domain.entity.Client;
import rs.raf.user_service.domain.entity.Company;
import rs.raf.user_service.exceptions.CompanyNotFoundException;
import rs.raf.user_service.exceptions.CompanyRegNumExistsException;
import rs.raf.user_service.exceptions.TaxIdAlreadyExistsException;
import rs.raf.user_service.domain.mapper.CompanyMapper;
import rs.raf.user_service.repository.ActivityCodeRepository;
import rs.raf.user_service.repository.ClientRepository;
import rs.raf.user_service.repository.CompanyRepository;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.stream.Collectors;

@Service
@AllArgsConstructor
public class CompanyService {

    private final CompanyRepository companyRepository;
    private final ClientRepository clientRepository;
    private final ActivityCodeRepository activityCodeRepository;

    public CompanyDto createCompany(CreateCompanyDto createCompanyDto) {
        Client client = clientRepository.findById(createCompanyDto.getMajorityOwner()).orElse(null);
        if (client == null)
            throw new NoSuchElementException("Owner not found with ID: " + createCompanyDto.getMajorityOwner());

        ActivityCode activityCode = activityCodeRepository.findById(createCompanyDto.getActivityCode()).orElse(null);
        if (activityCode == null)
            throw new NoSuchElementException("Activity code not found with ID: " + createCompanyDto.getActivityCode());
        if (companyRepository.findByRegistrationNumber(createCompanyDto.getRegistrationNumber()).isPresent())
            throw new CompanyRegNumExistsException(createCompanyDto.getRegistrationNumber());
        if (companyRepository.findByTaxId(createCompanyDto.getTaxId()).isPresent())

            throw new TaxIdAlreadyExistsException(createCompanyDto.getTaxId());
        Company company = new Company();
        company.setName(createCompanyDto.getName());
        company.setRegistrationNumber(createCompanyDto.getRegistrationNumber());
        company.setTaxId(createCompanyDto.getTaxId());
        company.setActivityCode(activityCode.getId());
        company.setAddress(createCompanyDto.getAddress());
        company.setMajorityOwner(client);

        return CompanyMapper.toDto(companyRepository.save(company));
    }

    public CompanyDto getCompanyById(Long id){
        Company company = companyRepository.findById(id)
                .orElseThrow(() -> new CompanyNotFoundException(id));
        return CompanyMapper.toDto(company);
    }

    public List<CompanyDto> getCompaniesForClientId(Long id){
        return companyRepository.findByMajorityOwner_Id(id).stream().map(CompanyMapper::toDto).collect(Collectors.toList());
    }


}

File: user-service-kotlin/src/main/kotlin/pack/userservicekotlin/service/CompanyService.kt
package pack.userservicekotlin.service

import arrow.core.Either
import arrow.core.left
import arrow.core.right
import org.springframework.stereotype.Service
import pack.userservicekotlin.arrow.CompanyServiceError
import pack.userservicekotlin.domain.dto.company.CompanyResponseDto
import pack.userservicekotlin.domain.dto.company.CreateCompanyDto
import pack.userservicekotlin.domain.entities.Company
import pack.userservicekotlin.domain.mapper.toDto
import pack.userservicekotlin.repository.ActivityCodeRepository
import pack.userservicekotlin.repository.ClientRepository
import pack.userservicekotlin.repository.CompanyRepository

@Service
class CompanyService(
    private val companyRepository: CompanyRepository,
    private val clientRepository: ClientRepository,
    private val activityCodeRepository: ActivityCodeRepository,
) {
    fun createCompany(createCompanyDto: CreateCompanyDto): Either<CompanyServiceError, CompanyResponseDto> {
        val ownerId = createCompanyDto.majorityOwner!!
        val activityCodeId = createCompanyDto.activityCode!!

        val client =
            clientRepository.findById(ownerId).orElse(null)
                ?: return CompanyServiceError.OwnerNotFound(ownerId).left()

        val activityCode =
            activityCodeRepository.findById(activityCodeId).orElse(null)
                ?: return CompanyServiceError.ActivityCodeNotFound(activityCodeId).left()

        // Uncomment these when validations are in place
//
//        companyRepository.findByRegistrationNumber(createCompanyDto.registrationNumber!!)?.let {
//            return CompanyServiceError.RegistrationNumberExists(createCompanyDto.registrationNumber).left()
//        }
//
//        companyRepository.findByTaxId(createCompanyDto.taxId!!)?.let {
//            return CompanyServiceError.TaxIdExists(createCompanyDto.taxId).left()
//        }

        val company =
            Company(
                name = createCompanyDto.name,
                registrationNumber = createCompanyDto.registrationNumber,
                taxId = createCompanyDto.taxId,
                activityCode = activityCode.id,
                address = createCompanyDto.address,
                majorityOwner = client,
            )

        return companyRepository.save(company).toDto()!!.right()
    }

    fun getCompanyById(id: Long): Either<CompanyServiceError, CompanyResponseDto> =
        companyRepository
            .findById(id)
            .map { it.toDto()!! }
            .orElse(null)
            ?.right()
            ?: CompanyServiceError.CompanyNotFound(id).left()

    fun getCompaniesForClientId(id: Long): List<CompanyResponseDto> = companyRepository.findByMajorityOwnerId(id).map { it.toDto()!! }
}
